Https client

ugandalf opened the pull request:
Resolves #3. Adds hyper-tls as a dependency for https connector support. In its current form exposes https client via `https_client()` function. Some tests included in `path.rs`. Schema itself was modified to fix compile errors and structs moved inside `Path` capability as they were only used there.

ErikMcClure commented on core/schema/http.capnp:
```diff
@@ -0,0 +1,48 @@
+
+interface Domain {
+  subdomain @0 (name :Text) -> (result :Domain);
+  path @1 (value :Text) -> (result :Path);
```
Please make this a `:List(Text)` and rename it "values" to reach parity with the path interface

ugandalf replied:
Did that in 4f05c1e934bad7a07fbd9a3851f7884c8d2e8844 and I consider it resolved.

ErikMcClure commented on core/tests/http_test.rs:
```diff
@@ -0,0 +1,368 @@
+    }
+    let _ = request.send().promise.await?.get()?.get_result()?;
+
+    ::core::result::Result::Err(::capnp::Error::unimplemented(
```
Please add a "finalize_domain" function to the domain interface that functions like the other finalize functions, and fill out the rest of this test accordingly.

ugandalf replied:
"finalize_domain" was done in d361c45614cf91a95b8e828bd4d7180c7b8f628c, tests added in following three commits, but I'm uncertain about how the edge cases should be handled. As it is, paths ending with .. and such are valid, but don't refer to parent directory. Trying to use http escapes results in path ending with "&period%3B&period%3B". Also unsure if that's how protocol relative url breakout would look like, so review would be appreciated. Otherwise I also consider it resolved.

aiverson reviewed (commented):
This test doesn't actually test what it says it does. This creates a new empty request to add things to the query string, then tests what query strings are in the newly created empty request, which is empty. It does not send anything to the provider and thus can't actually retrieve any data from it. The desired assurance is that there isn't a way to actually retrieve the already set query parameters from a path, to resist attempts to exfiltrate API tokens when handed an API url.

aiverson commented on core/tests/http_test.rs:
```diff
@@ -0,0 +1,548 @@
+    let mut request = finalized_client.query_request();
+    {
+        let values = request.get().get_values()?;
+        assert_eq!(values.len(), 0);
```
this test also doesn't assure anything useful

ugandalf replied:
Is the intention of that test to re-send existing request with changed data? Otherwise analogous to https://github.com/Fundament-Software/keystone/pull/10#discussion_r1134877641

aiverson replied:
So, what this test is doing is creating an empty request and then checking that the request is empty. The closest thing to that which makes any sense would be trying to retrieve the tests from the remote and verifying that it doesn't provide any. But that's not doable because there is deliberately no method to do so. So it should be treated like the other requirement of not being able to acquire something, and turned into a comment to make a human stop and reconsider since it isn't practical to automatically test it. Perhaps the test can be changed to verify some other piece of functionality about the query strings, or just removed entirely.

aiverson commented on core/tests/http_test.rs:
```diff
@@ -0,0 +1,548 @@
+        values_builder.reborrow().set(0, "/");
+    }
+    let err = request.send().promise.await;
+
```
this doesn't actually assert anything

aiverson replied:
This is presumably supposed to be checking if putting / on the path permits navigating up the tree, but it needs to assert that whatever it gets is an error from the "/" being an invalid component of a path, or that the result of the get is something other than the expected root item and instead something down the requested path.

aiverson commented on core/tests/http_test.rs:
```diff
@@ -0,0 +1,548 @@
+    let mut request = path_client.path_request();
+    {
+        let mut values_builder = request.get().init_values(1);
+        values_builder.reborrow().set(0, "&period;&period;");
```
this is an xml entity and not an http url escape code.

aiverson replied:
please use the url library's facilities for handling url encoding to verify the safety.

aiverson commented on core/src/http/path.rs:
```diff
@@ -0,0 +1,591 @@
+        assert_eq!(
+            request.send().promise.await.err().map(|e| e.description),
+            Some(
+                "Can't include GET in verb whitelist, because it's not in the original one"
```
this error doesn't necessarily make sense.

ugandalf commented on core/tests/http_test.rs:
```diff
@@ -0,0 +1,548 @@
+    // Attempting to get path
+    let mut request = path_client.path_request();
+    {
+        let values = request.get().get_values()?;
```
I don't know how to prove impossibility of retrieving previous components via a test, when we don't even have a request that retrieves existing path data. This test was written by Erik and since I don't know what needs to be tested, I left it as-is. Do you want me to remove that test altogether or do you have something in mind?

aiverson replied:
Yes, remove that part of the testing since there isn't a good place to actually test it. Instead, add comments to the capnp schema right next to where any methods that could exfiltrate data would be about how methods which permit the exfiltration need to not be added. That seems like about the best assurance we can do on that part for now.

ugandalf commented:
I addressed all the points, I think - url is first constructed with `url` library, then it's converted to string and then that is straight up fed to `hyper::Uri`. In the current state I'm not using `percent-encoding` library - **do you want me to?** I'm also very unsure about corectness / what the intended behavior should be when adding those manually - when manually printing (via `println!("Path: {}", url);` at line 79 of path.rs), I currently get for example paths like `https://httpbin.org/get/%252E%252E?` - `?` added when unnecessary, `%25` being a code for `%`.

Also before merging I think duplication of `lib.rs` and `main.rs` should be resolved, but `lib.rs` can't be simply removed.
